packages/eslint-plugin/src/rules/no-circular-import.ts

```diff
@@ -0,0 +1,181 @@
+import * as ts from 'typescript';
+
+import { createRule, getParserServices } from '../util';
+
+type Options = [];
+type MessageIds = 'noCircularImport';
+
+interface Edge {
+  filename: string;
+  specifier: string;
+}
+
+class Graph {
+  private graph = new Map<
+    string,
+    {
+      filename: string;
+      specifier: string;
+    }[]
+  >();
+
+  addEdge(start: string, edge: Edge): void {
+    if (this.graph.has(start)) {
+      this.graph.get(start)?.push(edge);
+    } else {
+      this.graph.set(start, [edge]);
+    }
+  }
+
+  hasEdge(name: string): boolean {
+    return this.graph.has(name);
+  }
+
+  getEdges(name: string): Edge[] {
+    return this.graph.get(name) ?? [];
+  }
+}
+//  imports “a.ts” and is imported from “b.ts”, resulting in a circular reference.
+export default createRule<Options, MessageIds>({
+  name: 'no-circular-import',
+  meta: {
+    docs: {
+      description:
+        'Disallow the use of import module that result in circular imports',
```

Comment:
```suggestion
        'Disallow a module from importing another module that transitively imports itself.',
```

`import/no-cycle` uses the description "Forbid a module from importing a module with a dependency path back to itself.". I think that's better.